### PR TITLE
feat!: Apply feemarket to native cosmos tx

### PR DESCRIPTION
**Pull request by yihuang**
## Description

Apply feemarket to native cosmos tx

- add tx extension option for user to input tip price
- apply feemarket's base fee to native tx
- fallback to default sdk logic when london hardfork not enabled

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/keeper/sdk_tx.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+// CheckSDKTxFee implements the `TxFeeChecker` for ante handler,
+// could be called in both checkTx and deliverTx modes.
+func (k Keeper) CheckSDKTxFee(ctx sdk.Context, tx sdk.Tx) (sdk.Coins, int64, error) {
+	params := k.GetParams(ctx)
+	denom := params.EvmDenom
+	ethCfg := params.ChainConfig.EthereumConfig(k.ChainID())
+
+	prioPriceCap := sdkmath.NewInt(0)
+	if hasExtOptsTx, ok := tx.(ante.HasExtensionOptionsTx); ok {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go:25]: When no extension, the tip fee will be default to `0`.

File: x/evm/keeper/sdk_tx.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+	sdkmath "cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/evmos/ethermint/x/evm/types"
+)
+
+var DefaultPriorityReduction = sdkmath.NewIntFromUint64(1000000)
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go:15]: Because the `priority` is a `int64` while token could be very big, similar to the rationale of `DefaultPowerReduction`.

[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go:15]: added comments:
```
// DefaultPriorityReduction is the default amount of price values required for 1 unit of priority.
// Because priority is `int64` while price is `big.Int`, it's necessary to scale down the range to keep it more pratical.
// The default value is the same as the `sdk.DefaultPowerReduction`.
```

File: x/evm/keeper/sdk_tx.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		}
+	}
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return nil, 0, fmt.Errorf("Tx must be a FeeTx")
+	}
+
+	gas := feeTx.GetGas()
+	feeCoins := feeTx.GetFee()
+	fee := feeCoins.AmountOf(denom)
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go]: done

File: x/evm/keeper/sdk_tx.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/evmos/ethermint/x/evm/types"
+)
+
+var DefaultPriorityReduction = sdkmath.NewIntFromUint64(1000000)
+
+// CheckSDKTxFee implements the `TxFeeChecker` for ante handler,
+// could be called in both checkTx and deliverTx modes.
+func (k Keeper) CheckSDKTxFee(ctx sdk.Context, tx sdk.Tx) (sdk.Coins, int64, error) {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go]: done

File: x/evm/keeper/sdk_tx.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fedekunze) [reply x/evm/keeper/sdk_tx.go:18]: let's use `var DefaultPriorityReduction = sdk.DefaultPowerReduction` instead

**Comment by fedekunze** [comment]
@yihuang can you fix the conflicts

**Comment by yihuang** [comment]
> @yihuang can you fix the conflicts

done, and extracted the evm tx priority part to separate PR, should do that one first: https://github.com/evmos/ethermint/pull/1214

File: x/evm/keeper/sdk_tx.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	if !ok {
+		return nil, 0, fmt.Errorf("Tx must be a FeeTx")
+	}
+
+	gas := feeTx.GetGas()
+	feeCoins := feeTx.GetFee()
+	fee := feeCoins.AmountOf(denom)
+	priceCap := fee.Quo(sdkmath.NewIntFromUint64(gas))
+
+	basePrice := sdkmath.NewIntFromBigInt(k.GetBaseFee(ctx, ethCfg))
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go:47]: I guess we should fallback to sdk default logic if `baseFee == nil`, need to make the function in cosmos-sdk public then.

[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go]: done

File: x/evm/keeper/sdk_tx.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go]: this var is moved to another PR.

File: CHANGELOG.md [hunk 6]
```diff
@@ -46,6 +46,11 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (ante) [#1176](https://github.com/evmos/ethermint/pull/1176) Fix invalid tx hashes; Remove `Size_` field and validate `Hash`/`From` fields in ante handler,
   recompute eth tx hashes in JSON-RPC APIs to fix old blocks.
 * (deps) [#1168](https://github.com/evmos/ethermint/pull/1168) Upgrade cosmos-sdk to v0.46.
+* (feemarket) [#1194](https://github.com/evmos/ethermint/pull/1194) Apply feemarket to native cosmos tx.
+
+### API Breaking
+
+* (ante) [#1214](https://github.com/evmos/ethermint/pull/1214) Set mempool priority to evm transactions.
```
Comment (yihuang) [inline CHANGELOG.md:53]:
```suggestion
```

File: app/simulation_test.go [hunk 7]
```diff
@@ -62,6 +63,32 @@ func fauxMerkleModeOpt(bapp *baseapp.BaseApp) {
 	bapp.SetFauxMerkleMode()
 }
 
+// NewSimApp disable feemarket on native tx, otherwise the cosmos-sdk simulation tests will fail.
+func NewSimApp(logger log.Logger, db dbm.DB) (*EthermintApp, error) {
+	encodingConfig := MakeEncodingConfig()
+	app := NewEthermintApp(logger, db, nil, false, map[int64]bool{}, DefaultNodeHome, simapp.FlagPeriodValue, MakeEncodingConfig(), simapp.EmptyAppOptions{}, fauxMerkleModeOpt)
```
Comment (yihuang) [inline app/simulation_test.go:69]:
```suggestion
	app := NewEthermintApp(logger, db, nil, false, map[int64]bool{}, DefaultNodeHome, simapp.FlagPeriodValue, encodingConfig, simapp.EmptyAppOptions{}, fauxMerkleModeOpt)
```

File: x/evm/keeper/sdk_tx.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	if !bigPriority.IsInt64() {
+		priority = math.MaxInt64
+	} else {
+		priority = bigPriority.Int64()
+	}
+	return
+}
+
+// checkTxFeeWithValidatorMinGasPrices implements the default fee logic, where the minimum price per
+// unit of gas is fixed and set by each validator, can the tx priority is computed from the gas price.
```
Comment (facs95) [inline x/evm/keeper/sdk_tx.go:70]: you meant, "and the tx.." ?

**Review by facs95 — APPROVED** [review]
I think this looks good so I will approve it but I will open a discussion. 

I would prefer `sdk_tx.go` logic to be in the Feemarket module instead of the EVM module. I understand that it needs the `LondonHardFork` param so it will be kind of tricky to have this in the FeeMarket so happy to leave it like this. I just think it will be the cleaner approach if there was a way to make it happen.

File: x/evm/keeper/sdk_tx.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go]: I changed it back, it seems [go linter](https://github.com/alexkohler/nakedret) suggest against named variables when function is not trivial.

File: x/evm/keeper/sdk_tx.go (hunk 8 above)
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go:70]: yes, changed

File: app/ante/ante.go [hunk 9]
```diff
@@ -45,6 +45,9 @@ func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
 				case "/ethermint.types.v1.ExtensionOptionsWeb3Tx":
 					// handle as normal Cosmos SDK tx, except signature is checked for EIP712 representation
 					anteHandler = newCosmosAnteHandlerEip712(options)
+				case "/ethermint.evm.v1.ExtensionOptionDynamicFeeTx":
```
Comment (fedekunze) [inline app/ante/ante.go:48]: the problem with using a tx extension is that I don't think wallets like Keplr  / Cosmostation will use it

File: x/evm/keeper/sdk_tx.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	sdkmath "cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/evmos/ethermint/x/evm/types"
+)
+
+// CheckSDKTxFee implements the `TxFeeChecker` for ante handler,
+// could be called in both checkTx and deliverTx modes.
```
Comment (fedekunze) [inline x/evm/keeper/sdk_tx.go:16]: imo this name is not very descriptive. Also, the logic is not well documented here, what does this function does? The comment only mentions where it's used

File: x/feemarket/client/cli/query.go [hunk 11]
```diff
@@ -36,27 +30,20 @@ func GetQueryCmd() *cobra.Command {
 // GetBlockGasCmd queries the gas used in a block
 func GetBlockGasCmd() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "block-gas [height]",
+		Use:   "block-gas",
 		Short: "Get the block gas used at a given block height",
 		Long: `Get the block gas used at a given block height.
 If the height is not provided, it will use the latest height from context`,
-		Args: cobra.RangeArgs(0, 1),
+		Args: cobra.NoArgs,
```
Comment (fedekunze) [inline x/feemarket/client/cli/query.go:37]: these changes are unrelated to the PR, please split it into a separate PR

File: x/evm/keeper/sdk_tx.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
+package keeper
```
Comment (fedekunze) [inline x/evm/keeper/sdk_tx.go:1]: can we move this to the `app/ante/` directory?

File: x/evm/keeper/sdk_tx.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	params := k.GetParams(ctx)
+	denom := params.EvmDenom
+	ethCfg := params.ChainConfig.EthereumConfig(k.ChainID())
+	baseFee := k.GetBaseFee(ctx, ethCfg)
+
+	if baseFee == nil {
+		// london hardfork is not enabled: fallback to min-gas-prices logic
+		return checkTxFeeWithValidatorMinGasPrices(ctx, tx)
```
Comment (fedekunze) [inline x/evm/keeper/sdk_tx.go:30]: not covered by tests

**Review by fedekunze — CHANGES_REQUESTED** [review]
Thank for the PR. How do you see wallets implementing the extension? If we don't use it by default most wallets won't use this logic. I'd change the default to be the fee market to cosmos txs and the extension for the old logic.

Also, this PR is missing unit tests

File: app/ante/ante.go (hunk 9 above)
Comment (yihuang) [reply app/ante/ante.go:48]: yeah, it's a matter of standards, there was a discussion before of providing a feemarket module in cosmos-sdk, but we didn't reach a consensus.
Maybe better to make the `max_prio_price` default to max int, the default behavior would be more intuitive, the user pay full `gas_price`, and gets priority proportional to `gas_price - base_fee`.

Comment (fedekunze) [reply app/ante/ante.go:48]: correct, but we'd need the post tx handler for that (as I mentioned in the other issue) to get the refund after the tx is processed

File: proto/ethermint/evm/v1/tx.proto [hunk 14]
```diff
@@ -172,3 +172,11 @@ message MsgEthereumTxResponse {
   // gas consumed by the transaction
   uint64 gas_used = 5;
 }
+
+// cosmos-sdk tx extension option
+message ExtensionOptionDynamicFeeTx {
+  string max_priority_price = 1 [
```
[the remark this answers is not in the document]
Comment (yihuang) [reply proto/ethermint/evm/v1/tx.proto:178]: do you mean put it next to the `ExtensionOptionsEthereumTx`?

[the remark this answers is not in the document]
Comment (yihuang) [reply proto/ethermint/evm/v1/tx.proto]: done, moved to `proto/ethermint/types/dynamic_fee.proto`, side with the `web3.proto`.

File: app/ante/ante.go (hunk 9 above)
Comment (yihuang) [reply app/ante/ante.go]: opened issue: https://github.com/evmos/ethermint/issues/1231

File: x/evm/keeper/sdk_tx.go (hunk 10 above)
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go]: done, added more comments.

File: x/feemarket/client/cli/query.go (hunk 11 above)
Comment (yihuang) [reply x/feemarket/client/cli/query.go:37]: done, extracted to https://github.com/evmos/ethermint/pull/1230

File: x/evm/keeper/sdk_tx.go (hunk 12 above)
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go]: done

File: x/evm/keeper/sdk_tx.go (hunk 13 above)
Comment (yihuang) [reply x/evm/keeper/sdk_tx.go]: added unit test.

**Comment by yihuang** [comment]
> Thank for the PR. How do you see wallets implementing the extension? If we don't use it by default most wallets won't use this logic. I'd change the default to be the fee market to cosmos txs and the extension for the old logic.

- I guess It'll take some time for keplr to support but we can try, hopefully creating some standard.
- Apps that build tx by themselves are able to use it.
- The default behavior when the extension option is missing should be intuitive.

> 
> Also, this PR is missing unit tests

added

**Review by fedekunze — APPROVED** [review]
ACK. I did a minor refactor to make the code more maintainable. Great work!

**Comment by yihuang** [comment]
> ACK. I did a minor refactor to make the code more maintainable. Great work!

the gofmt warning can't be reproduced locally, any ideas @facs95 ?
